aslam_cv_common/test/test-stl-helpers.cc

```diff
@@ -41,25 +41,36 @@ TEST(StlHelpers, EraseIndicesAligned) {
 }
 
 TEST(StlHelpers, DrawRandom) {
-  Aligned<std::vector, Eigen::Vector3i> test_vector;
-  test_vector.push_back(Eigen::Vector3i::Constant(0));
-  test_vector.push_back(Eigen::Vector3i::Constant(1));
-  test_vector.push_back(Eigen::Vector3i::Constant(2));
-  test_vector.push_back(Eigen::Vector3i::Constant(3));
-
-  Aligned<std::vector, Eigen::Vector3i> output;
+  std::vector<int> test_vector = {0, 1, 2, 3, 4, 5};
+  std::vector<int> output;
 
-  const size_t kNum = 3;
-  aslam::common::drawNRandomElements(kNum, test_vector, &output, true);
-  EXPECT_EQ(output.size(), kNum);
+  constexpr size_t kNumToPick = 3u;
+  aslam::common::drawNRandomElements(kNumToPick, test_vector, &output, true);
+  EXPECT_EQ(output.size(), kNumToPick);
 
   output.clear();
-  aslam::common::drawNRandomElements(kNum, test_vector, &output, false);
-  EXPECT_EQ(output.size(), kNum);
+  aslam::common::drawNRandomElements(kNumToPick, test_vector, &output, false);
+  EXPECT_EQ(output.size(), kNumToPick);
 
   output.clear();
-  aslam::common::drawNRandomElements(kNum, test_vector, &output);
-  EXPECT_EQ(output.size(), kNum);
+  aslam::common::drawNRandomElements(kNumToPick, test_vector, &output);
+  EXPECT_EQ(output.size(), kNumToPick);
+
+  // Fail when input and output are the same, because this is not allowed.
+  EXPECT_DEATH(aslam::common::drawNRandomElements(
+      kNumToPick, test_vector, &test_vector), "");
+
+  // Make sure that if we pick a single index 1000 times, every index
+  // (i.e., 0 to 5) is at least picked once.
+  std::unordered_set<int> indices_encountered;
+  const size_t kNumTrials = 1000u;
+  for (size_t i = 0u; i < kNumTrials; ++i) {
+    constexpr size_t kNumToPick = 1u;
+    aslam::common::drawNRandomElements(kNumToPick, test_vector, &output, false);
+    ASSERT_EQ(output.size(), kNumToPick);
+    indices_encountered.emplace(output[0u]);
+  }
+  EXPECT_EQ(indices_encountered, std::unordered_set<int>({0, 1, 2, 3, 4, 5}));
```

Comment:
This still has a small chance to fail, so it's a bit suboptimal. Or maybe we can ignore the small error chance, the chance that one number isn't picked seems to be low enough: 0.8^1000=1.2302e-97

Reply:
Yes, I thought the same! I actually made it now pick 3 instead of only one in each draw, that lowers the change to 9.332636e-302 ;-)
Fixing the seed doesn't help here, because in that case it will repeatedly pick the same index.
I don't see a nice way to solve this, and I would also argue that the chance of failing are sufficiently small. 
But if you have a suggestion how to somehow make this deterministic and at the same time make sure that the whole range of indices are picked from, then please let me know. Otherwise I'd keep it like this for now.